### DISCUSSION TITLE
How does kong route to the specified backend service based on url matching?

### DISCUSSION THREAD

**Discussion by zyiqian** [Kong/kong#12394] [Help] [answered]
Hi everyone..

### Problem
When I want to route the traffic with the url "open-api.cn:10443/cloud/connect" to the open-api-stable backend service, when the url is "open-api.cn:10443/cloud/connect?sn= 123abc" traffic is routed to the open-api-canary service. How do I implement it through kong?
Does kong support routing based on the query parameter sn=123abc?

### My KONG setup
- kong:3.1
- kong-ingress-controller:2.8

does anyone help me in this issue?

**Comment by nowNick** [comment]
Hi @zyiqian !

I'll be happy to help. Have you tried experimenting with Kong Expressions router? You could configure your route to be matched based on query parameters. Here's more about Expressions Router and it's available matching fields: https://docs.konghq.com/gateway/3.5.x/key-concepts/routes/expressions/#matching-fields

Remember that you have to enable expressions based routing. More info here: https://docs.konghq.com/gateway/latest/key-concepts/routes/expressions/

Does this answer your question?

**Reply by zyiqian** [reply]
Hi @nowNick ,

Thank you very much for your answer. I deployed kong using DB-less mode, and it seems that routing cannot be configured using expressions.

**Reply by chronolaw** [reply]
DB-less mode can work with expressions router perfectly, you can try:
```
KONG_DATABASE=off \
KONG_ROUTER_FLAVOR=expressions \
kong start
```

**Reply by zyiqian** [reply]
Hi @chronolaw ,

Thanks for your quick reply, I configured kong according to your suggestion and it works properly. But I still can't match the fields using "http.queries".

Configure "http.headers" and it works
```
expression: '(http.path == "/") && (http.host == "test-xxxx.com") && (http.headers.x_my_header == "example")'
```

Configure "http.queries" it doesn't work
```
expression: '(http.path == "/") && (http.host == "test-xxxx.com") && (http.queries.x_my_queries == "example")'
```

Returns the following error
```
│ proxy nginx: [error] init_by_lua error: /usr/local/share/lua/5.1/kong/init.lua:608: error parsing declarative config file /kong_dbless/kong.yml:                                                                
│ proxy in 'services':                                                                                                                                                                                            
│ proxy   - in entry 4 of 'services':                                                                                                                                                                             
│ proxy     in 'routes':                                                                                                                                                                                          
│ proxy       - in entry 1 of 'routes':                                                                                                                                                                           
│ proxy         in '@entity':                                                                                                                                                                                     
│ proxy           - in entry 1 of '@entity': Router Expression failed validation: Unknown LHS field
```

Any good suggestions?

**Comment by chronolaw** [comment, marked as answer]
Kong gateway supports http queries only after 3.4, could you upgrade it and try again?
